In this project, we'll be cleaning some data from ebay Kleinanzeigen, a classified section of the German eBay website. I know nothing about German, this should be fun.

Data Dictionary  
dateCrawled - When this ad was first crawled. All field-values are taken from this date.  
name - Name of the car.  
seller - Whether the seller is private or a dealer.  
offerType - The type of listing.  
price - The price on the ad to sell the car.  
abtest - Whether the listing is included in an A/B test.  
vehicleType - The vehicle Type.  
yearOfRegistration - The year in which the car was first registered.  
gearbox - The transmission type.  
powerPS - The power of the car in PS.  
model - The car model name.  
kilometer - How many kilometers the car has driven.  
monthOfRegistration - The month in which the car was first registered.  
fuelType - What type of fuel the car uses.  
brand - The brand of the car.  
notRepairedDamage - If the car has a damage which is not yet repaired.  
dateCreated - The date on which the eBay listing was created.  
nrOfPictures - The number of pictures in the ad.  
postalCode - The postal code for the location of the vehicle.  
lastSeenOnline - When the crawler saw this ad last online.  

In [1]:
import pandas as pd
import numpy as np

In [2]:
autos = pd.read_csv("autos.csv")

UnicodeDecodeError: 'utf-8' codec can't decode byte 0xdc in position 23: invalid continuation byte

As it turns out, this data won't load using UTF-8. Trying other encodings...

In [93]:
autos = pd.read_csv("autos.csv", encoding='Latin-1')

In [94]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_

In [95]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

We've got some rows that don't have all fields filled in. Will need to determine what to do with those.

In [96]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

For the most part, the words can be roughly translated to English. Not sure what Angebot is. Might need to look that up. Some of the fields (price, odometer) have units or special characters that we'll need to remove. Should probably figure out if monthofRegistration is 0 or 1 based.

Now let's take the headings and make them a bit more readable/meaningful. In this case, it was easier to print out the column names and adjust them in a text editor before feeding them back in.

In [97]:
autos_cols = autos.columns

print(autos_cols)

autos_cols = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'number_of_pictures', 'postal_code',
       'last_seen']

autos.columns = autos_cols

autos.head()

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  number_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00                   0   
1         bmw              nein  2016-04-04 00:00:00                   0   
2  volkswagen              nein  2016-03-26 00:00:00                   0   
3       smart              nein  2016-03-12 00:00:00                   0   
4        ford              nein  2016-04-01 00:00:00                   0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In [98]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-04-02 11:37:04  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  number_of_pictures   postal_code  \
count                 50000             50000.0  50000.000000   
unique                   76                 NaN           NaN   
top     2016-04-03 00:00:00                 NaN           NaN   
freq                   1946                 NaN           NaN   
mean                    NaN                 0.0  50813.627300   
std                     NaN                 0.0  25779.747957   
min                     NaN                 0.0   1067.000000   
25%                     NaN                 0.0  30451.000000   
50%                     NaN                 0.0  49577.000000   
75%                     NaN                 0.0  71540.000000   
max                     NaN                 0.0  99998.000000   

                  last_seen  
count                 50000  
unique                39481  
top     2016-04-07 06:17:27  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

Based on the output, our to-do list looks like the following:  
price - Remove \$ signs  
price - Lots of \$0 values. This needs research.  
registration_year - minimum 1000, maximum 9999. This needs more cleaning.  
odometer - remove km  
registration_month - minimum 0, maximum 12. This needs more cleaning.  
number_of_pictures - Data is all zero. We can remove.  
vehicle_type - Count is only 44905. Why?  
gearbox - Count is only 47320. Why?  
model - Count is only 47242. Why?



Let's start with the odometer column. Since all of the values appear to be in km, we can change the column heading to include that information and, hopefully, convert the data to integers.

In [99]:
autos = autos.rename(columns={'odometer': 'odometer_km'})

In [100]:
autos['odometer_km'] = autos['odometer_km'].str.replace('km', '')
autos['odometer_km'] = autos['odometer_km'].str.replace(',','')
autos['odometer_km'] = autos['odometer_km'].astype(int)

In [101]:
autos.dtypes

date_crawled          object
name                  object
seller                object
offer_type            object
price                 object
ab_test               object
vehicle_type          object
registration_year      int64
gearbox               object
power_ps               int64
model                 object
odometer_km            int64
registration_month     int64
fuel_type             object
brand                 object
unrepaired_damage     object
ad_created            object
number_of_pictures     int64
postal_code            int64
last_seen             object
dtype: object

We've successfully cleaned up that column and converted it to integers. Let's do the same for price

In [102]:
autos['price'] = autos['price'].str.replace('$', '')
autos['price'] = autos['price'].str.replace(',','')
autos['price'] = autos['price'].astype(int)

In [103]:
autos.dtypes

date_crawled          object
name                  object
seller                object
offer_type            object
price                  int64
ab_test               object
vehicle_type          object
registration_year      int64
gearbox               object
power_ps               int64
model                 object
odometer_km            int64
registration_month     int64
fuel_type             object
brand                 object
unrepaired_damage     object
ad_created            object
number_of_pictures     int64
postal_code            int64
last_seen             object
dtype: object

Now let's take a look at the price data some more. On first glance, there are lots of cars with extremely low prices (<\$50) which seems odd. Maybe someone's just trying to dump a junker? Maybe it's a car PART listed incorrectly? It's probably safe to say a car should cost more than \$50. Also, shout out to everyone trying to sell their car for \$1 ala The Price is Right.

In [104]:
autos['price'].value_counts().sort_index(ascending=True)

0           1421
1            156
2              3
3              1
5              2
8              1
9              1
10             7
11             2
12             3
13             2
14             1
15             2
17             3
18             1
20             4
25             5
29             1
30             7
35             1
40             6
45             4
47             1
49             4
50            49
55             2
59             1
60             9
65             5
66             1
            ... 
151990         1
155000         1
163500         1
163991         1
169000         1
169999         1
175000         1
180000         1
190000         1
194000         1
197000         1
198000         1
220000         1
250000         1
259000         1
265000         1
295000         1
299000         1
345000         1
350000         1
999990         1
999999         2
1234566        1
1300000        1
3890000        1
10000000       1
11111111       2
12345678      

In [105]:
autos = autos[autos['price'] > 50]

In [106]:
autos['price'].value_counts().sort_index(ascending=True)

55            2
59            1
60            9
65            5
66            1
70           10
75            5
79            1
80           15
89            1
90            5
99           19
100         134
110           3
111           2
115           2
117           1
120          39
122           1
125           8
129           1
130          15
135           1
139           1
140           9
145           2
149           7
150         224
156           2
160           8
           ... 
151990        1
155000        1
163500        1
163991        1
169000        1
169999        1
175000        1
180000        1
190000        1
194000        1
197000        1
198000        1
220000        1
250000        1
259000        1
265000        1
295000        1
299000        1
345000        1
350000        1
999990        1
999999        2
1234566       1
1300000       1
3890000       1
10000000      1
11111111      2
12345678      3
27322222      1
99999999      1
Name: price, Length: 233

In [107]:
autos[autos['price'] > 500000]

date_crawled                                               name  \
514    2016-03-17 09:53:08                   Ford_Focus_Turnier_1.6_16V_Style   
2897   2016-03-12 21:50:57   Escort_MK_1_Hundeknochen_zum_umbauen_auf_RS_2000   
7814   2016-04-04 11:53:31                                        Ferrari_F40   
11137  2016-03-29 23:52:57  suche_maserati_3200_gt_Zustand_unwichtig_laufe...   
22947  2016-03-22 12:54:19                         Bmw_530d_zum_ausschlachten   
24384  2016-03-21 13:57:51                            Schlachte_Golf_3_gt_tdi   
27371  2016-03-09 15:45:47                                         Fiat_Punto   
37585  2016-03-29 11:38:54                                Volkswagen_Jetta_GT   
39377  2016-03-08 23:53:51                        Tausche_volvo_v40_gegen_van   
39705  2016-03-22 14:58:27                        Tausch_gegen_gleichwertiges   
42221  2016-03-08 20:39:05                                  Leasinguebernahme   
43049  2016-03-21 19:53:52                                      2_VW_Busse_T3   
47598  2016-03-31 18:56:54  Opel_Vectra_B_1_6i_16V_Facelift_Tuning_Showcar...   
47634  2016-04-04 21:25:21                                        Ferrari_FXX   

       seller offer_type     price  ab_test vehicle_type  registration_year  \
514    privat    Angebot    999999     test        kombi               2009   
2897   privat    Angebot  11111111     test    limousine               1973   
7814   privat    Angebot   1300000  control        coupe               1992   
11137  privat    Angebot  10000000  control        coupe               1960   
22947  privat    Angebot   1234566  control        kombi               1999   
24384  privat    Angebot  11111111     test          NaN               1995   
27371  privat    Angebot  12345678  control          NaN               2017   
37585  privat    Angebot    999990     test    limousine               1985   
39377  privat    Angebot  12345678  control          NaN               2018   
39705  privat    Angebot  99999999  control    limousine               1999   
42221  privat    Angebot  27322222  control    limousine               2014   
43049  privat    Angebot    999999     test          bus               1981   
47598  privat    Angebot  12345678  control    limousine               2001   
47634  privat    Angebot   3890000     test        coupe               2006   

         gearbox  power_ps        model  odometer_km  registration_month  \
514      manuell       101        focus       125000                   4   
2897     manuell        48       escort        50000                   3   
7814         NaN         0          NaN        50000                  12   
11137    manuell       368          NaN       100000                   1   
22947  automatik       190          NaN       150000                   2   
24384        NaN         0          NaN       150000                   0   
27371        NaN        95        punto       150000                   0   
37585    manuell       111        jetta       150000                  12   
39377    manuell        95          v40       150000                   6   
39705  automatik       224     s_klasse       150000                   9   
42221    manuell       163           c4        40000                   2   
43049    manuell        70  transporter       150000                   1   
47598    manuell       101       vectra       150000                   3   
47634        NaN       799          NaN         5000                   7   

      fuel_type           brand unrepaired_damage           ad_created  \
514      benzin            ford              nein  2016-03-17 00:00:00   
2897     benzin            ford              nein  2016-03-12 00:00:00   
7814        NaN  sonstige_autos              nein  2016-04-04 00:00:00   
11137    benzin  sonstige_autos              nein  2016-03-29 00:00:00   
22947    diesel             bmw               NaN  2016-03-22 00:00:00   
24384       NaN      volksw

There are some dollar amounts that clearly look out of place, but there are also some dollar amounts that COULD be valid. Let's get rid of the really egregious ones for the moment. The others might filter out as we deal with other columns.

In [108]:
autos.drop(autos[autos['price'] == 12345678].index, inplace = True) 
autos.drop(autos[autos['price'] == 99999999].index, inplace = True) 
autos.drop(autos[autos['price'] == 11111111].index, inplace = True) 

In [109]:
autos['odometer_km'].value_counts().sort_index(ascending=True)

5000        773
10000       248
20000       757
30000       777
40000       816
50000      1012
60000      1153
70000      1214
80000      1413
90000      1733
100000     2104
125000     5046
150000    31260
Name: odometer_km, dtype: int64

The odometer data is way too clean, but nothing seems horribly out of place so we can keep it.

In [110]:
autos['date_crawled']

0        2016-03-26 17:47:46
1        2016-04-04 13:38:56
2        2016-03-26 18:57:24
3        2016-03-12 16:58:10
4        2016-04-01 14:38:50
5        2016-03-21 13:47:45
6        2016-03-20 17:55:21
7        2016-03-16 18:55:19
8        2016-03-22 16:51:34
9        2016-03-16 13:47:02
10       2016-03-15 01:41:36
11       2016-03-16 18:45:34
12       2016-03-31 19:48:22
13       2016-03-23 10:48:32
14       2016-03-23 11:50:46
15       2016-04-01 12:06:20
16       2016-03-16 14:59:02
17       2016-03-29 11:46:22
18       2016-03-26 19:57:44
19       2016-03-17 13:36:21
20       2016-03-05 19:57:31
21       2016-03-06 19:07:10
22       2016-03-28 20:50:54
23       2016-03-10 19:55:34
24       2016-04-03 11:57:02
25       2016-03-21 21:56:18
26       2016-04-03 22:46:28
28       2016-03-19 21:56:19
29       2016-04-02 12:45:44
30       2016-03-14 11:47:31
                ...         
49968    2016-04-01 17:49:15
49969    2016-03-17 18:49:02
49970    2016-03-21 22:47:37
49971    2016-

In [111]:
autos['date_crawled'] = autos['date_crawled'].str[:10]
autos['ad_created'] = autos['ad_created'].str[:10]
autos['last_seen'] = autos['last_seen'].str[:10]

In [112]:
autos['date_crawled'].value_counts().sort_index(ascending=True)

2016-03-05    1225
2016-03-06     680
2016-03-07    1739
2016-03-08    1607
2016-03-09    1596
2016-03-10    1557
2016-03-11    1573
2016-03-12    1786
2016-03-13     757
2016-03-14    1770
2016-03-15    1656
2016-03-16    1423
2016-03-17    1524
2016-03-18     622
2016-03-19    1677
2016-03-20    1825
2016-03-21    1801
2016-03-22    1591
2016-03-23    1560
2016-03-24    1423
2016-03-25    1523
2016-03-26    1559
2016-03-27    1502
2016-03-28    1687
2016-03-29    1650
2016-03-30    1627
2016-03-31    1539
2016-04-01    1630
2016-04-02    1718
2016-04-03    1862
2016-04-04    1766
2016-04-05     631
2016-04-06     153
2016-04-07      67
Name: date_crawled, dtype: int64

In [113]:
autos['ad_created'].value_counts().sort_index(ascending=True)

2015-06-11       1
2015-08-10       1
2015-09-09       1
2015-11-10       1
2015-12-05       1
2015-12-30       1
2016-01-03       1
2016-01-07       1
2016-01-10       2
2016-01-13       1
2016-01-14       1
2016-01-16       1
2016-01-22       1
2016-01-27       3
2016-01-29       1
2016-02-01       1
2016-02-02       2
2016-02-05       2
2016-02-07       1
2016-02-08       1
2016-02-09       1
2016-02-11       1
2016-02-12       2
2016-02-14       2
2016-02-16       1
2016-02-17       1
2016-02-18       2
2016-02-19       3
2016-02-20       2
2016-02-21       3
              ... 
2016-03-09    1600
2016-03-10    1543
2016-03-11    1589
2016-03-12    1777
2016-03-13     823
2016-03-14    1704
2016-03-15    1643
2016-03-16    1447
2016-03-17    1508
2016-03-18     655
2016-03-19    1623
2016-03-20    1828
2016-03-21    1812
2016-03-22    1582
2016-03-23    1552
2016-03-24    1420
2016-03-25    1529
2016-03-26    1562
2016-03-27    1498
2016-03-28    1692
2016-03-29    1648
2016-03-30  

In [114]:
autos['last_seen'].value_counts().sort_index(ascending=True)

2016-03-05       52
2016-03-06      209
2016-03-07      262
2016-03-08      354
2016-03-09      464
2016-03-10      513
2016-03-11      599
2016-03-12     1147
2016-03-13      429
2016-03-14      611
2016-03-15      767
2016-03-16      794
2016-03-17     1356
2016-03-18      354
2016-03-19      762
2016-03-20      997
2016-03-21      994
2016-03-22     1032
2016-03-23      898
2016-03-24      956
2016-03-25      926
2016-03-26      806
2016-03-27      751
2016-03-28     1007
2016-03-29     1079
2016-03-30     1194
2016-03-31     1150
2016-04-01     1105
2016-04-02     1204
2016-04-03     1213
2016-04-04     1184
2016-04-05     6040
2016-04-06    10714
2016-04-07     6383
Name: last_seen, dtype: int64

Nothing really looks out of place with the dates now that we've removed the times to be able to summarize them easier. Let's move on to the registration year.

In [115]:
autos['registration_year'].value_counts().sort_index(ascending=False)

9999       3
9000       1
8888       1
6200       1
5911       1
5000       3
4800       1
4500       1
4100       1
2800       1
2019       2
2018     468
2017    1385
2016    1205
2015     382
2014     663
2013     801
2012    1309
2011    1620
2010    1589
2009    2081
2008    2212
2007    2274
2006    2670
2005    2917
2004    2699
2003    2695
2002    2478
2001    2629
2000    3114
        ... 
1964      12
1963       8
1962       4
1961       6
1960      23
1959       6
1958       4
1957       2
1956       4
1955       2
1954       2
1953       1
1952       1
1951       2
1950       1
1948       1
1943       1
1941       2
1939       1
1938       1
1937       4
1934       2
1931       1
1929       1
1927       1
1910       2
1800       2
1111       1
1001       1
1000       1
Name: registration_year, Length: 95, dtype: int64

Well....that's not right. It would stand to reason that any cars registered on or before 1800 were...horses and buggies?  Let's get rid of those as well as anything registered after 2019 to start with. The more recent years might require additional research.

In [116]:
autos = autos[(autos['registration_year'] <= 2019) & (autos['registration_year'] >= 1910)]

In [117]:
autos['registration_year'].value_counts().sort_index(ascending=False)

2019       2
2018     468
2017    1385
2016    1205
2015     382
2014     663
2013     801
2012    1309
2011    1620
2010    1589
2009    2081
2008    2212
2007    2274
2006    2670
2005    2917
2004    2699
2003    2695
2002    2478
2001    2629
2000    3114
1999    2882
1998    2348
1997    1935
1996    1361
1995    1198
1994     629
1993     421
1992     368
1991     338
1990     334
        ... 
1968      26
1967      26
1966      22
1965      17
1964      12
1963       8
1962       4
1961       6
1960      23
1959       6
1958       4
1957       2
1956       4
1955       2
1954       2
1953       1
1952       1
1951       2
1950       1
1948       1
1943       1
1941       2
1939       1
1938       1
1937       4
1934       2
1931       1
1929       1
1927       1
1910       2
Name: registration_year, Length: 81, dtype: int64

In [118]:
autos[autos['registration_year'] > 2016]

date_crawled                                               name  seller  \
10      2016-03-15                       VW_Golf_Tuning_in_siber/grau  privat   
65      2016-04-04                      Ford_Fiesta_zum_ausschlachten  privat   
68      2016-04-03  Mini_cooper_s_clubman_/vollausstattung_/_Navi/...  privat   
84      2016-03-27                                     Renault_twingo  privat   
113     2016-04-03                              Golf_4_Anfaenger_auto  privat   
164     2016-03-13  Opel_Meriva__nur_76000_Km__unfallfrei__scheckh...  privat   
197     2016-04-05                              VW_Polo_9N_an_Bastler  privat   
253     2016-03-27             Ford_mondeo_Gas_anlage_mit_TÜV_04.2017  privat   
348     2016-03-17  VW_Beetle_1.8Turbo_mit_Vollausstattung_und_seh...  privat   
390     2016-03-25   Fiat_Bertone_X_1_9__X_1/9__X19__X_19__X1_9__X_19  privat   
438     2016-03-10                VW_Golf_5_V_1.6_2004_Klima_Tempomat  privat   
454     2016-03-25    Ford_Fiesta_Servo_D3_Kat_86€_Steuern_65.000_tkm  privat   
457     2016-03-15  Mercedes_Benz_W203_C200_Kombi_Kompressor_c_kla...  privat   
477     2016-03-31                                         Citroën_C4  privat   
545     2016-03-22                                      Bmw_x5_/__E53  privat   
548     2016-03-10            Peugeot_206_1.9_Hdi_4_Tueren_Tuev_02.17  privat   
557     2016-03-16                           Smart_Fortwo_Passion_MHD  privat   
568     2016-03-18             PRIMA_ANFÄNGERAUTO_smart_fortwo_cabrio  privat   
577     2016-04-03                                  Volkswagen_Golf_3  privat   
582     2016-03-31                                    V10_TDI_TOUAREG  privat   
654     2016-04-01              Ich_verkaufe_meinen_gepflegten_daewoo  privat   
740     2016-04-02                            Seat_Leon_zu_verkaufen!  privat   
754     2016-03-26                         Ford_Ka_Kleinwagen_BJ_2008  privat   
758     2016-03-12                            Honda_Civic_mit_TÜV_!!!  privat   
765     2016-03-26               BMW_3er_Limousine_mit_1_Jahr_TÜV!!!!  privat   
802     2016-03-19                 Lada_mit_wenig_km_neuem_Tuev_bj_08  privat   
871     2016-04-04                                   Citroen_C2_60_PS  privat   
889     2016-03-12                          Verkaufe_Opel_Corsa_C_1_2  privat   
910     2016-03-15                         Renault_Kangoo_bj.2007_LPG  privat   
918     2016-03-15                              BMW_114i_Sondermodell  privat   
...            ...                                                ...     ...   
49075   2016-03-18           VW_Golf_2_GT_TÜV_neu_GTi_G60_16V_Special  privat   
49076   2016-03-25                             Omega_B_2.0_16v_136_Ps  privat   
49108   2016-03-14                  TDI_20_L_DIESEL140_PS..AUTOMATICK  privat   
49130   2016-03-09                                      Renault_Modus  privat   
49154   2016-03-24                    Auto_mit_TÜV_bis_September_2017  privat   
49178   2016-03-24       Mazda_323_F_TÜV__neu_2te_Hand_viele_Neuteile  privat   
49185   2016-03-28              Alfa_Romeo_156_Sportwagon_1_8_T_Spark  privat   
49218   2016-04-01          Ford_Fiesta_1.3_4/5_Tuerer_16_Monate_Tuev  privat   
49245   2016-03-26                         Audi_A3_1.8T_tausch_moegl.  privat   
49259   2016-03-19                  Volkswagen_Lupo_TÜV_06.2017_Klima  privat   
49261   2016-03-10                                        Toyota_aygo  privat   
49262   2016-03-26                                        Opel_vectra  privat   
49266   2016-03-10  Toyota_Corolla_Verso_7_Sitzer_104.000km_2_2_Di...  privat   
49343   2016-03-12                                   bmw_320i_touring  privat   
49347   2016-03-21                         Honda_civic_mit_TÜV_3/2017  privat   
49354   2016-04-05                        Bmw_e39_523i_mit_neuem_Tuev  privat   
49389   2016-03-10                                         Astra_G_CC  privat   
49411   2016-03-28                   

The registration years after 2016 are probably incorrect as this data came from 2016. Let's get rid of those as well.

In [119]:
autos = autos[autos['registration_year'] <= 2016]

In [123]:
autos['registration_year'].value_counts().sort_index(ascending=True)

1910       2
1927       1
1929       1
1931       1
1934       2
1937       4
1938       1
1939       1
1941       2
1943       1
1948       1
1950       1
1951       2
1952       1
1953       1
1954       2
1955       2
1956       4
1957       2
1958       4
1959       6
1960      23
1961       6
1962       4
1963       8
1964      12
1965      17
1966      22
1967      26
1968      26
        ... 
1987      72
1988     134
1989     171
1990     334
1991     338
1992     368
1993     421
1994     629
1995    1198
1996    1361
1997    1935
1998    2348
1999    2882
2000    3114
2001    2629
2002    2478
2003    2695
2004    2699
2005    2917
2006    2670
2007    2274
2008    2212
2009    2081
2010    1589
2011    1620
2012    1309
2013     801
2014     663
2015     382
2016    1205
Name: registration_year, Length: 78, dtype: int64

That's probably as good as we're going to get with the registration_year field. Let's move on to the brand and see if we can pick out any particular subset of data we'd like to review.

In [128]:
autos['brand'].value_counts().sort_index(ascending=True)

alfa_romeo         309
audi              4024
bmw               5110
chevrolet          264
chrysler           164
citroen            654
dacia              123
daewoo              69
daihatsu           117
fiat              1194
ford              3245
honda              365
hyundai            465
jaguar              72
jeep               106
kia                329
lada                27
lancia              49
land_rover          98
mazda              709
mercedes_benz     4484
mini               408
mitsubishi         380
nissan             712
opel              4981
peugeot           1388
porsche            281
renault           2187
rover               62
saab                77
seat               848
skoda              763
smart              658
sonstige_autos     446
subaru              98
suzuki             275
toyota             593
trabant             64
volkswagen        9811
volvo              423
Name: brand, dtype: int64

In [131]:
autos['brand'].value_counts().sort_values(ascending=False)

volkswagen        9811
bmw               5110
opel              4981
mercedes_benz     4484
audi              4024
ford              3245
renault           2187
peugeot           1388
fiat              1194
seat               848
skoda              763
nissan             712
mazda              709
smart              658
citroen            654
toyota             593
hyundai            465
sonstige_autos     446
volvo              423
mini               408
mitsubishi         380
honda              365
kia                329
alfa_romeo         309
porsche            281
suzuki             275
chevrolet          264
chrysler           164
dacia              123
daihatsu           117
jeep               106
subaru              98
land_rover          98
saab                77
jaguar              72
daewoo              69
trabant             64
rover               62
lancia              49
lada                27
Name: brand, dtype: int64

This is a place where normalizing might come in really handy. What if we just wanted to look at any brands that had more than 1% of the postings?

In [132]:
autos['brand'].value_counts(normalize=True).sort_values(ascending=False)

volkswagen        0.211298
bmw               0.110053
opel              0.107275
mercedes_benz     0.096571
audi              0.086664
ford              0.069887
renault           0.047101
peugeot           0.029893
fiat              0.025715
seat              0.018263
skoda             0.016433
nissan            0.015334
mazda             0.015270
smart             0.014171
citroen           0.014085
toyota            0.012771
hyundai           0.010015
sonstige_autos    0.009605
volvo             0.009110
mini              0.008787
mitsubishi        0.008184
honda             0.007861
kia               0.007086
alfa_romeo        0.006655
porsche           0.006052
suzuki            0.005923
chevrolet         0.005686
chrysler          0.003532
dacia             0.002649
daihatsu          0.002520
jeep              0.002283
subaru            0.002111
land_rover        0.002111
saab              0.001658
jaguar            0.001551
daewoo            0.001486
trabant           0.001378
r

In this case, Hyundai would be our cutoff point, although I'm going to keep sonstige_autos for the moment because I've never heard of it and I'm genuinely curoius given that I've heard of so many other brands on the list. We will do our filtering using a dictionary generation loop.

In [137]:
auto_filter = autos['brand'].value_counts(normalize=True).sort_values(ascending=False).index[:18]

In [138]:
auto_filter

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat', 'skoda', 'nissan', 'mazda', 'smart',
       'citroen', 'toyota', 'hyundai', 'sonstige_autos'],
      dtype='object')

In [170]:
brand_mean_prices = {}

for brand in auto_filter:
    brand_mean_prices[brand] = autos[autos['brand'] == brand]['price'].mean()


In [175]:
brand_mean_prices = (sorted(brand_mean_prices.items(), 
                           key = lambda kv: kv[1], 
                           reverse = True))
brand_mean_prices

[('sonstige_autos', 46728.32735426009),
 ('citroen', 45556.23547400611),
 ('audi', 9376.088469184891),
 ('mercedes_benz', 8664.980151650312),
 ('bmw', 8618.393346379647),
 ('skoda', 6393.018348623853),
 ('volkswagen', 5634.23850779737),
 ('hyundai', 5399.651612903226),
 ('toyota', 5167.091062394604),
 ('nissan', 4750.063202247191),
 ('seat', 4423.116745283019),
 ('mazda', 4112.596614950635),
 ('ford', 4078.2748844375965),
 ('smart', 3596.40273556231),
 ('peugeot', 3105.0886167146973),
 ('opel', 2999.6338084721942),
 ('fiat', 2820.7747068676717),
 ('renault', 2490.5665294924556)]

Well....that's interesting. Sonstige Autos comes out on top. I'm guessing it's a boutique company...maybe something you'd see on Instagram?

Some Googling leads us in a completely different direction.  

German: sonstige  
English:  other  
    
Well...that's not as exciting as I had hoped for. Given that they could come from any number of places, let's just get rid of them.

In [176]:
brand_mean_prices.pop(0)

('sonstige_autos', 46728.32735426009)

In [179]:
brand_mean_prices

[('citroen', 45556.23547400611),
 ('audi', 9376.088469184891),
 ('mercedes_benz', 8664.980151650312),
 ('bmw', 8618.393346379647),
 ('skoda', 6393.018348623853),
 ('volkswagen', 5634.23850779737),
 ('hyundai', 5399.651612903226),
 ('toyota', 5167.091062394604),
 ('nissan', 4750.063202247191),
 ('seat', 4423.116745283019),
 ('mazda', 4112.596614950635),
 ('ford', 4078.2748844375965),
 ('smart', 3596.40273556231),
 ('peugeot', 3105.0886167146973),
 ('opel', 2999.6338084721942),
 ('fiat', 2820.7747068676717),
 ('renault', 2490.5665294924556)]

Why do we have a jump from 45k to 9k in average price?  Let's look at the Citroen cars and see if something's amiss.

In [186]:
autos[autos['brand'] == 'citroen']['price'].sort_values(ascending=False)

42221    27322222
48944       33000
6476        25000
21720       22900
11315       21111
23431       21000
20065       20900
43891       20650
23063       17900
47501       16000
10826       15823
49970       15800
45873       15400
21488       15000
40765       14999
21954       14890
15418       14800
9128        14500
25033       14400
22735       14350
2874        14000
7208        12900
14682       12750
45073       12290
4884        11980
43570       11899
10560       11750
14586       11500
14843       11450
32715       11300
           ...   
8115          350
26184         350
36573         350
35168         350
32512         340
19687         333
9268          300
36306         300
24759         300
5925          300
6328          300
33690         300
43714         280
43079         280
38692         250
20493         250
22140         250
36452         250
15083         250
6431          222
16785         200
32303         200
27133         200
18691         200
19181     

As it turns out, a $27M car probably doesn't fit in given the rest of the prices. Let's just get rid of any cars over \$500k and start over.

In [205]:
autos = pd.read_csv("autos.csv", encoding='Latin-1')

autos_cols = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'number_of_pictures', 'postal_code',
       'last_seen']

autos.columns = autos_cols

autos = autos.rename(columns={'odometer': 'odometer_km'})

autos['odometer_km'] = autos['odometer_km'].str.replace('km', '')
autos['odometer_km'] = autos['odometer_km'].str.replace(',','')
autos['odometer_km'] = autos['odometer_km'].astype(int)

autos['price'] = autos['price'].str.replace('$', '')
autos['price'] = autos['price'].str.replace(',','')
autos['price'] = autos['price'].astype(int)

autos.drop(autos[autos['price'] > 500000].index, inplace = True) 

autos['date_crawled'] = autos['date_crawled'].str[:10]
autos['ad_created'] = autos['ad_created'].str[:10]
autos['last_seen'] = autos['last_seen'].str[:10]

autos = autos[(autos['registration_year'] <= 2019) & (autos['registration_year'] >= 1910)]

autos = autos[autos['registration_year'] <= 2016]

auto_filter = autos['brand'].value_counts(normalize=True).sort_values(ascending=False).index[:18]

brand_mean_prices = {}

for brand in auto_filter:
    brand_mean_prices[brand] = autos[autos['brand'] == brand]['price'].mean()
    
del brand_mean_prices['sonstige_autos'] 

brand_mean_prices


{'audi': 9093.65003615329,
 'bmw': 8102.536248343744,
 'citroen': 3699.935628742515,
 'fiat': 2711.8011272141707,
 'ford': 3652.095223880597,
 'hyundai': 5308.53911205074,
 'mazda': 4010.7716643741405,
 'mercedes_benz': 8485.239571958942,
 'nissan': 4664.891034482758,
 'opel': 2876.716403542549,
 'peugeot': 3039.4682651622,
 'renault': 2395.4164467897976,
 'seat': 4296.492554410081,
 'skoda': 6334.91948051948,
 'smart': 3542.706586826347,
 'toyota': 5115.33388981636,
 'volkswagen': 5231.081983308787}

This looks much better. Now let's take a look at the average mileage for these brands to see if there's a link in mileage and price.

In [208]:
brand_mean_mileage = {}

for brand in auto_filter:
    brand_mean_mileage[brand] = autos[autos['brand'] == brand]['odometer_km'].mean()
    
del brand_mean_mileage['sonstige_autos'] 

brand_mean_mileage

{'audi': 129287.78018799711,
 'bmw': 132431.38368351315,
 'citroen': 119580.8383233533,
 'fiat': 116553.94524959743,
 'ford': 124068.65671641791,
 'hyundai': 106511.62790697675,
 'mazda': 124745.5295735901,
 'mercedes_benz': 130856.0821139987,
 'nissan': 118572.41379310345,
 'opel': 129223.14208702349,
 'peugeot': 127136.81241184767,
 'renault': 128183.81706244503,
 'seat': 121563.57388316152,
 'skoda': 110954.54545454546,
 'smart': 99595.80838323354,
 'toyota': 115709.51585976628,
 'volkswagen': 128724.10407461954}

In [209]:
bmp = pd.Series(brand_mean_prices)
bmm = pd.Series(brand_mean_mileage)

In [210]:
bmp

audi             9093.650036
bmw              8102.536248
citroen          3699.935629
fiat             2711.801127
ford             3652.095224
hyundai          5308.539112
mazda            4010.771664
mercedes_benz    8485.239572
nissan           4664.891034
opel             2876.716404
peugeot          3039.468265
renault          2395.416447
seat             4296.492554
skoda            6334.919481
smart            3542.706587
toyota           5115.333890
volkswagen       5231.081983
dtype: float64

In [211]:
bmm

audi             129287.780188
bmw              132431.383684
citroen          119580.838323
fiat             116553.945250
ford             124068.656716
hyundai          106511.627907
mazda            124745.529574
mercedes_benz    130856.082114
nissan           118572.413793
opel             129223.142087
peugeot          127136.812412
renault          128183.817062
seat             121563.573883
skoda            110954.545455
smart             99595.808383
toyota           115709.515860
volkswagen       128724.104075
dtype: float64

In [212]:
df_bmp_bmm = (pd.DataFrame(bmp, 
                            columns = ['mean_price']))

# add the 2nd series as a new column to the dataframe
df_bmp_bmm['mean_km'] = bmm
(df_bmp_bmm.sort_values(by = 'mean_price', 
                             ascending = False))

mean_price        mean_km
audi           9093.650036  129287.780188
mercedes_benz  8485.239572  130856.082114
bmw            8102.536248  132431.383684
skoda          6334.919481  110954.545455
hyundai        5308.539112  106511.627907
volkswagen     5231.081983  128724.104075
toyota         5115.333890  115709.515860
nissan         4664.891034  118572.413793
seat           4296.492554  121563.573883
mazda          4010.771664  124745.529574
citroen        3699.935629  119580.838323
ford           3652.095224  124068.656716
smart          3542.706587   99595.808383
peugeot        3039.468265  127136.812412
opel           2876.716404  129223.142087
fiat           2711.801127  116553.945250
renault        2395.416447  128183.817062

In [215]:
df_bmp_bmm['price_per_km'] = df_bmp_bmm['mean_price']/df_bmp_bmm['mean_km']

In [216]:
df_bmp_bmm

mean_price        mean_km  price_per_km
audi           9093.650036  129287.780188      0.070337
bmw            8102.536248  132431.383684      0.061183
citroen        3699.935629  119580.838323      0.030941
fiat           2711.801127  116553.945250      0.023266
ford           3652.095224  124068.656716      0.029436
hyundai        5308.539112  106511.627907      0.049840
mazda          4010.771664  124745.529574      0.032152
mercedes_benz  8485.239572  130856.082114      0.064844
nissan         4664.891034  118572.413793      0.039342
opel           2876.716404  129223.142087      0.022262
peugeot        3039.468265  127136.812412      0.023907
renault        2395.416447  128183.817062      0.018687
seat           4296.492554  121563.573883      0.035344
skoda          6334.919481  110954.545455      0.057095
smart          3542.706587   99595.808383      0.035571
toyota         5115.333890  115709.515860      0.044208
volkswagen     5231.081983  128724.104075      0.040638

I think what we're seeing here is expected. Higher value cars retain that value, even if they have more miles on them. Aside from that, I'm not sure this comparison generates any value as the numbers don't hold much of a trend outside of Audi, BMW, and Mercedes Benz.

Next Steps...

Data cleaning next steps:  
* Identify categorical data that uses german words, translate them and map the values to their english counterparts  
* Convert the dates to be uniform numeric data, so "2016-03-21" becomes the integer 20160321.  
* See if there are particular keywords in the name column that you can extract as new columns  
  
Analysis next steps: 
* Find the most common brand/model combinations  
* Split the odometer_km into groups, and use aggregation to see if average prices follows any patterns based on the mileage.  
* How much cheaper are cars with damage than their non-damaged counterparts?